internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -44,15 +44,19 @@ var (
 //
 // It implements the [resolver.Resolver] interface.
 type delegatingResolver struct {
-	target         resolver.Target     // parsed target URI to be resolved
-	cc             resolver.ClientConn // gRPC ClientConn
-	targetResolver resolver.Resolver   // resolver for the target URI, based on its scheme
-	proxyResolver  resolver.Resolver   // resolver for the proxy URI; nil if no proxy is configured
-	proxyURL       *url.URL            // proxy URL, derived from proxy environment and target
+	target   resolver.Target     // parsed target URI to be resolved
+	cc       resolver.ClientConn // gRPC ClientConn
+	proxyURL *url.URL            // proxy URL, derived from proxy environment and target
 
 	mu                  sync.Mutex         // protects all the fields below
 	targetResolverState *resolver.State    // state of the target resolver
 	proxyAddrs          []resolver.Address // resolved proxy addresses; empty if no proxy is configured
+
+	// childMu serializes calls into child resolvers. It also protects access to
+	// the following fields.
+	childMu        sync.Mutex
```

Comment:
Can we not share `mu` for this?  Why not?

Reply:
`childMu` is being used to serialize calls into the child resolvers, similar to the [serailzer queue in resolver_wrapper.go](https://github.com/grpc/grpc-go/blob/a51009d1d7074ee1efcd323578064cbe44ef87e5/resolver_wrapper.go#L98-L124).

If we re-use the existing mutex, we would need to hold `mu` while calling `childResolver.ResolveNow()`. If this calls `cc.UpdateState()` synchronously, it will try to re-acquire `mu` and deadlock.

* `mu` is used to serialize `resolver -> channel` calls.
* `childMu` is used to serialize `channel -> resolver` calls.